dask/dataframe/io/io.py

```diff
@@ -750,5 +755,245 @@ def sorted_division_locations(seq, npartitions=None, chunksize=None):
     return values, positions
 
 
+class _PackedArgCallable(DataFrameIOFunction):
+    """Packed-argument wrapper for DataFrameIOFunction
+
+    This is a private helper class for ``from_map``. This class
+    ensures that packed positional arguments will be expanded
+    before the underlying function (``func``) is called. This class
+    also handles optional metadata enforcement and column projection
+    (when ``func`` satisfies the ``DataFrameIOFunction`` protocol).
+    """
+
+    def __init__(
+        self,
+        func,
+        args=None,
+        kwargs=None,
+        meta=None,
+        packed=False,
+        enforce_metadata=False,
+    ):
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.meta = meta
+        self.enforce_metadata = enforce_metadata
+        self.packed = packed
+        self.is_dataframe_io_func = isinstance(self.func, DataFrameIOFunction)
+
+    @property
+    def columns(self):
+        if self.is_dataframe_io_func:
+            return self.func.columns
+        return None
+
+    def project_columns(self, columns):
+        if self.is_dataframe_io_func:
+            return _PackedArgCallable(
+                self.func.project_columns(columns),
+                args=self.args,
+                kwargs=self.kwargs,
+                meta=self.meta,
+                packed=self.packed,
+                enforce_metadata=self.enforce_metadata,
+            )
+        return self
+
+    def __call__(self, packed_arg):
+        if not self.packed:
+            packed_arg = [packed_arg]
+        if self.enforce_metadata:
+            return apply_and_enforce(
+                *packed_arg,
+                *(self.args or []),
+                _func=self.func,
+                _meta=self.meta,
+                **(self.kwargs or {}),
+            )
+        return self.func(
+            *packed_arg,
+            *(self.args or []),
+            **(self.kwargs or {}),
+        )
+
+
+@insert_meta_param_description
+def from_map(
+    func,
+    *iterables,
+    args=None,
+    meta=None,
+    divisions=None,
+    label=None,
+    token=None,
+    enforce_metadata=True,
+    **kwargs,
+):
+    """Create a DataFrame collection from a custom function map
+
+    Parameters
+    ----------
+    func : callable
+        Function used to create each partition. If ``func`` satisfies the
+        ``DataFrameIOFunction`` protocol, column projection will be enabled.
+    *iterables : Iterable objects
+        Iterable objects to map to each output partition. all iterables must
```

Comment:
typo:
```suggestion
        Iterable objects to map to each output partition. All iterables must
```
Also I was initially a little confused about the hierarchy of this arg. My current understanding is that iterables is a list of arbitratry len where each item within must have len = npartitions. So an example of an iterable would be a list of filenames right? I think this is what the description says, but it wasn't immediately obvious to me :shrug: